README

This is the dx operating system, version 0.1.2, released 2012-02-11



* WHAT IS THIS?
	A small OS for the x86.  See dx/doc/vision.txt.


* CURRENT STATUS
	Still under active development.  Not useful yet.  Basic micro kernel,
	some user mode libraries, couple of user executables.  Nothing else.
	Should boot cleanly from either floppy disk(!) or CD-ROM image.


* BUILD INSTRUCTIONS
	See the documentation under dx/doc/build_process.


* RELEASE NOTES and CHANGES FROM PREVIOUS VERSIONS
	See dx/doc/release_notes.txt.


* HARDWARE REQUIREMENTS
	See dx/doc/hardware_support.txt.


* ONLINE
	This project is officially hosted at https://bitbucket.org/dgaur/dx.
	See also http://www.dangaur.com/portfolio/dx.html.


* COPYRIGHT + LICENSE
	All code is distributed under the standard two-clause BSD license; and
	copyright (c) 2005 - 2011, Dan Gaur; except:

	* Doug Lea's malloc.c implementation, v2.8.5, from
	  http://g.oswego.edu/dl/html/malloc.html.  See dx/src/lib/libc/malloc.c.
	  Public domain.

	* The "freely distributable libm" math library, v5.3, from
	  http://www.netlib.org/fdlibm or http://www.validlab.com/software.
	  See dx/src/lib/libm.

	* LuaJIT, v1.1.7, from www.luajit.org.  See dx/src/lib/liblua.  Released
	  under MIT license.

	* David Gay's gdtoa library, from http://www.netlib.org/fp.  See
	  dx/src/lib/libgdtoa.





doc/release_notes.txt

* DX RELEASE NOTES *



* PENDING-RELEASE:

* Version 0.1.2, released 2012-02-11:
	Rearranged the ramdisk layout and boot sequence.  Moving toward separate
	"interpreter + script" model.


* Version 0.1.1, released 2012-01-26:
	Rudimentary file-system support; initial support for reading files from
	ramdisk.  User shell is now a standalone lua script.  Various associated
	bug fixes.


* Version 0.1.0, released 2011-12-11:
	Initial lua integration.  No changes in externally-visible functionality,
	but liblua is now viable and the user shell is built with lua.


* Version 0.0.7, released 2011-11-19:
	Various improvements to the build and release process.  No functional
	changes.


* Version 0.0.6, released 2011-11-09:
	Significant additions to libc, still aimed at porting Lua.  Incorporated
	and updated some third-party sources (libm, libgdtoa) for the same.  Couple
	bug fixes.  No major changes or additions in externally-visible
	functionality.


* Version 0.0.5, released 2011-08-20:
	Expanded and improved the libc implementation, mostly aimed at porting the
    Lua VM.  Some minor cleanup to the ramdisk.  No major changes or additions
	in (external) functionality.


* Version 0.0.4, released 2009-12-06:
	Feature/functionality release.  Improved infrastructure for drivers running
	in user space.  Keyboard driver is much improved, now supports SHIFT, CAPS
	LOCK, etc.  Added some basic kernel statistics for monitoring/debugging.
	Minor fixes to the VGA driver.  Some small performance improvements.
	Other small fixes.


* Version 0.0.3, released 2009-11-04:
	Maintenance and bug-fix release.  Various bug fixes, code cleanup and
	minor optimizations.  Documentation updates.  No major changes in
	functionality.


* Version 0.0.2, released 2009-10-14:
	Maintenance release.  Various build improvements + fixes.  Added support
	for recent versions of gcc + binutils.  Cleaned up doxygen output.
	Various updates to the build documentation.  No major code changes.


* Version 0.0.1a, released 2009-08-17:
	Minor fixes to the release process itself.  No actual changes to the OS
	logic.


* Version 0.0.1, released 2009-08-16:
	Initial public release on kenai.com.  Basic micro kernel; some user mode
	libraries; couple of user mode executables.  Sole interface is a very
	simple shell.  No other functionality.  Not useful yet.

	Appears to be stable(?) on basic 32-bit x86 UP platforms.  Includes support
	for basic VGA (text mode only); and (incomplete, very limited) support for
	PS/2 keyboards.  The debug kernel writes output to COM1, 9600 8N1.



src/inc/errno.h
//
// errno.h
//

#ifndef _ERRNO_H
#define _ERRNO_H

extern
int errno;


//
// Standard error codes.  EDOM, EILSEQ and ERANGE are defined by C99.  All
// other values are defined by POSIX/SUS
//
#define E2BIG			1	// Argument list too long.
#define EACCES			2	// Permission denied.
#define EADDRINUSE		3	// Address in use.
#define EADDRNOTAVAIL	4	// Address not available.
#define EAFNOSUPPORT	5	// Address family not supported.
#define EAGAIN			6	// Resource unavailable, try again
#define EALREADY		7	// Connection already in progress.
#define EBADF			8	// Bad file descriptor.
#define EBADMSG			9	// Bad message.
#define EBUSY			10	// Device or resource busy.
#define ECANCELED		11	// Operation canceled.
#define ECHILD			12	// No child processes.
#define ECONNABORTED	13	// Connection aborted.
#define ECONNREFUSED	14	// Connection refused.
#define ECONNRESET		15	// Connection reset.
#define EDEADLK			16	// Resource deadlock would occur.
#define EDESTADDRREQ	17	// Destination address required.
#define EDOM			18	// Mathematics argument out of domain of function.
#define EDQUOT			19	// Reserved.
#define EEXIST			20	// File exists.
#define EFAULT			21	// Bad address.
#define EFBIG			22	// File too large.
#define EHOSTUNREACH	23	// Host is unreachable.
#define EIDRM			24	// Identifier removed.
#define EILSEQ			25	// Illegal byte sequence.
#define EINPROGRESS		26	// Operation in progress.
#define EINTR			27	// Interrupted function.
#define EINVAL			28	// Invalid argument.
#define EIO				29	// I/O error.
#define EISCONN			30	// Socket is connected.
#define EISDIR			31	// Is a directory.
#define ELOOP			32	// Too many levels of symbolic links.
#define EMFILE			33	// Too many open files.
#define EMLINK			34	// Too many links.
#define EMSGSIZE		35	// Message too large.
#define EMULTIHOP		36	// Reserved.
#define ENAMETOOLONG	37	// Filename too long.
#define ENETDOWN		38	// Network is down.
#define ENETRESET		39	// Connection aborted by network.
#define ENETUNREACH		40	// Network unreachable.
#define ENFILE			41	// Too many files open in system.
#define ENOBUFS			42	// No buffer space available.
#define ENODATA			43	// No message is available on the STREAM
#define ENODEV			44	// No such device.
#define ENOENT			45	// No such file or directory.
#define ENOEXEC			46	// Executable file format error.
#define ENOLCK			47	// No locks available.
#define ENOLINK			48	// Reserved.
#define ENOMEM			49	// Not enough space.
#define ENOMSG			50	// No message of the desired type.
#define ENOPROTOOPT		51	// Protocol not available.
#define ENOSPC			52	// No space left on device.
#define ENOSR			53	// No STREAM resources.
#define ENOSTR			54	// Not a STREAM.
#define ENOSYS			55	// Function not supported.
#define ENOTCONN		56	// The socket is not connected.
#define ENOTDIR			57	// Not a directory.
#define ENOTEMPTY		58	// Directory not empty.
#define ENOTSOCK		59	// Not a socket.
#define ENOTSUP			60	// Not supported.
#define ENOTTY			61	// Inappropriate I/O control operation.
#define ENXIO			62	// No such device or address.
#define EOPNOTSUPP		63	// Operation not supported on socket.
#define EOVERFLOW		64	// Value too large to be stored in data type.
#define EPERM			65	// Operation not permitted.
#define EPIPE			66	// Broken pipe.
#define EPROTO			67	// Protocol error.
#define EPROTONOSUPPORT	68	// Protocol not supported.
#define EPROTOTYPE		69	// Protocol wrong type for socket.
#define ERANGE			70	// Result too large.
#define EROFS			71	// Read-only file system.
#define ESPIPE			72	// Invalid seek.
#define ESRCH			73	// No such process.
#define ESTALE			74	// Reserved.
#define ETIME			75	// Stream ioctl() timeout.
#define ETIMEDOUT		76	// Connection timed out.
#define ETXTBSY			77	// Text file busy.
#define EWOULDBLOCK		78	// Operation would block
#define EXDEV			79	// Cross-device link. 


#endif
